alibi/utils/distributed.py

```diff
@@ -217,3 +224,203 @@ def _chunk(values, chunksize):
+        A list containing the results for each batch, in the order that the batch was submitted to the parallel pool.
+        It may contain:
+            - `np.ndarray` objects (single-output predictor)
+            - lists of `np.ndarray` objects (multi-output predictors)
+
+    Returns
+    -------
+    concatenated
+        A list containing the concatenated results for all the batches.
+    """
+    if isinstance(ordered_result[0], np.ndarray):
+        return np.concatenate(ordered_result, axis=0)
+
+    # concatenate explanations for every class
+    n_classes = len(ordered_result[0])
+    to_concatenate = [list(zip(*ordered_result))[idx] for idx in range(n_classes)]
+    concatenated = [np.concatenate(arrays, axis=0) for arrays in to_concatenate]
+    return concatenated
+
+
+def invert_permutation(p: list):
+    """
+    Inverts a permutation.
+
+    Parameters:
+    -----------
+    p
+        Some permutation of 0, 1, ..., len(p)-1. Returns an array s, where s[i] gives the index of i in p.
+
+    Returns
+    -------
+    s
+        `s[i]` gives the index of `i` in `p`.
+    """
+
+    s = np.empty_like(p)
+    s[p] = np.arange(len(p))
+    return s
+
+
+class DistributedExplainer:
+    """
+    A class that orchestrates the execution of the execution of a batch of explanations in parallel.
+    """
+    if RAY_INSTALLED:
+        import ray
+        ray = ray
+
+    def __init__(self, distributed_opts, cls, init_args, init_kwargs):
+        if not RAY_INSTALLED:
+            raise ModuleNotFoundError("Module requires ray to be installed. pip install alibi[ray] ")
+
+        self.n_jobs = distributed_opts['n_cpus']
+        self.batch_size = distributed_opts['batch_size']
+        self.algorithm = distributed_opts['algorithm']
+        self.target_fn = globals()[f"{distributed_opts['algorithm']}_target_fn"]
+        try:
+            self.post_process_fcn = globals()[f"{distributed_opts['algorithm']}_postprocess_fn"]
+        except KeyError:
+            self.post_process_fcn = None
+
+        self.explainer = cls
+        self.explainer_args = init_args
+        self.explainer_kwargs = init_kwargs
+
+        if not DistributedExplainer.ray.is_initialized():
+            DistributedExplainer.ray.init()
+
+        self.pool = self.create_parallel_pool()
+
+    def __getattr__(self, item):
+        """
+        Access to actor attributes. Should be used to retrieve only state that is shared by all actors in the pool.
+        """
+        actor = self.pool._idle_actors[0]
+        return actor.return_attribute.remote(item)
+
+    def create_parallel_pool(self):
+        """
+        Creates a pool of actors that can execute explanations in parallel.
+        """
+
+        handles = [DistributedExplainer.ray.remote(self.explainer) for _ in range(self.n_jobs)]
+        workers = [handle.remote(*self.explainer_args, **self.explainer_kwargs) for handle in handles]
+        return DistributedExplainer.ray.util.ActorPool(workers)
```

Comment:
Is `ActorPool` now part of `ray` itself?

Reply:
Yes, same code as in our library.